app/services/recovery_code_generator.rb

```diff
@@ -0,0 +1,28 @@
+class RecoveryCodeGenerator
+  def initialize(user, length: 16)
+    @user = user
+    @length = length
+  end
+
+  def create
+    user.update!(recovery_code: hashed_code)
+
+    raw_recovery_code
+  end
+
+  private
+
+  attr_reader :length, :user
+
+  def hashed_code
+    Devise::Encryptor.digest(User, raw_recovery_code)
+  end
+
+  def raw_recovery_code
+    @raw_recovery_code ||= SecureRandom.hex(recovery_code_length / 2)
```

Comment:
It doesn't have to be, but it could lead to confusion if someone sets it to an odd number, but the code length is even. I'll add a comment to application.yml